# Deeplearning - Anees Ahmad - 2020/03/08

# 8 Introduction to deep learning for computer vision
- convolutional neural networks
  - convnets
  - used universally in computer vision applications
  - image-classification problems
  - small training datasets

---

## 8.1 Introduction to convnets
- a basic convnet
  - a stack of Conv2D and MaxPooling2D layers.
  - convnet takes as input tensors of shape `(image_height, image_width,image_channels)`

- build the model using the Functional API

In [1]:
# Listing 8.1 Instantiating a small convnet
from tensorflow import keras 
from tensorflow.keras import layers
# define the input shape
# as we are dealing with MNIST data we know it is a 28*28 pixels graysacale image
inputs = keras.Input(shape=(28, 28, 1))

# a convent is stacks of Conv2D and MaxPooling2D layers
# filter is actually the nodes/channels, karnel_size is the weight
# same layer can be created using sequential class as follows
  # model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
# pool size defines the factor with which it scale down
  # model.add(layers.MaxPooling2D((2, 2)))
x = layers.Conv2D(filters=32, kernel_size=3, activation="relu")(inputs)
x = layers.MaxPooling2D(pool_size=2)(x)

x = layers.Conv2D(filters=64, kernel_size=3, activation="relu")(x)
x = layers.MaxPooling2D(pool_size=2)(x)

x = layers.Conv2D(filters=128, kernel_size=3, activation="relu")(x)

- Output of Conv2D and MaxPooling2D layer
  - rank-3 tensor of shape `(height, width, channels)`

In [2]:
# next we have stacks of Dense layer, whihc actually takes 1D tensor as input
# need to flattern the output of last Conv28 Layer
x = layers.Flatten()(x)
outputs = layers.Dense(10, activation="softmax")(x)
model = keras.Model(inputs=inputs, outputs=outputs) 

In [3]:
# Listing 8.2 Displaying the model’s summary
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 28, 28, 1)]       0         
                                                                 
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 128)         73856 

In [4]:
# Listing 8.3 Training the convnet on MNIST images
from tensorflow.keras.datasets import mnist
 
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype("float32") / 255
model.compile(
    optimizer="rmsprop",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
938/938 [==============================] - 10s 9ms/step - loss: 0.1553 - accuracy: 0.9526
Epoch 2/5
938/938 [==============================] - 7s 8ms/step - loss: 0.0436 - accuracy: 0.9866
Epoch 3/5
938/938 [==============================] - 7s 8ms/step - loss: 0.0298 - accuracy: 0.9909
Epoch 4/5
938/938 [==============================] - 7s 8ms/step - loss: 0.0227 - accuracy: 0.9929
Epoch 5/5
938/938 [==============================] - 7s 8ms/step - loss: 0.0176 - accuracy: 0.9944


In [5]:
# Listing 8.4 Evaluating the convnet
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"Test accuracy: {test_acc:.3f}")

313/313 [==============================] - 1s 4ms/step - loss: 0.0315 - accuracy: 0.9900
Test accuracy: 0.990


- With out convents we have an accuracy of 97.8%
- With convents we have accuracy of 99.1%

---